GEC documentation (#57)
ALLUOS/ALLUOS.github.io: _posts/2021-02-01-NLP.md

image: pic08.jpg
---

## Contents

1. [Motivation](#motivation)
2. [Goal and scope](#goal-and-scope)
3. [Tools and Models](#tools-and-models)
	- [LanguageTool](#languagetool)
	- [GECToR](#gector)
	- [Fine-tuning and Datasets](#fine-tuning-and-datasets)
4. [Implementation](#implementation)
	- [Grammatical Error Detection Pipeline](#grammatical-error-detection-pipeline)
	- [Grammar Feedback and Passing the Task](#grammar-feedback-and-passing-the-task)
5. [Outlook and Limitations](#outlook-and-limitations)
6. [References](#references)

## Motivation

> Learning to correctly apply grammatical structures in written language is an important and often difficult task for English learners. Therefore, during this semester, grammatical error correction (GEC) was introduced in Escapeling in order to help learners improve their grammar skills. More specifically, state-of-the-art [NLP-tools for GEC](#tools-and-models) were employed in the [discussion task](link).

Reviewer: ```suggestion
Learning to correctly apply grammatical structures in written language is an important and often difficult task for English learners. Therefore, during this semester, grammatical error correction (GEC) was introduced in Escapeling in order to help learners improve their grammar skills. More specifically, state-of-the-art [NLP-tools for GEC](#tools-and-models) were employed in the [discussion task]({{ "" | absolute_url }}/2021/03/02/discussion.html).
```

> The discussion task offers a suitable framework for actively learning and practicing grammar because in this task, users actively produce language, as opposed to focusing on language comprehension in [other tasks](link). To that end, we include diverse feedback on the users’ contributions both after each round of the discussion and at the end of the task.

Reviewer: ```suggestion
The discussion task offers a suitable framework for actively learning and practicing grammar because in this task, users actively produce language, as opposed to focusing on language comprehension in [other tasks]({{ "" | absolute_url }}/#three). To that end, we include diverse feedback on the users’ contributions both after each round of the discussion and at the end of the task.
```

> Another reason for adding GEC to the discussion task is the improvement of the completion metric for this task. That is, in line with the [escape-room](link to storytelling) framing of the app, users’ performance is linked to getting pieces of the code in order to complete the mission. Previously, the performance was only evaluated quantitatively, i.e., by setting a threshold on the minimal number of words users need to produce in order to pass the task. We deem that adding a qualitative evaluation of the contributions to the completion metric improves the accuracy and is advantageous from a learning experience perspective. Therefore, we integrate a score computed via our GEC tools into the updated [completion metric](Implementation) of the discussion task. Finally, our integration of GEC tools can potentially be of use in other parts of the app in the future.

Reviewer: ```suggestion
Another reason for adding GEC to the discussion task is the improvement of the completion metric for this task. That is, in line with the [escape-room]({{ "" | absolute_url }}/2021/01/03/storytelling.html) framing of the app, users’ performance is linked to getting pieces of the code in order to complete the mission. Previously, the performance was only evaluated quantitatively, i.e., by setting a threshold on the minimal number of words users need to produce in order to pass the task. We deem that adding a qualitative evaluation of the contributions to the completion metric improves the accuracy and is advantageous from a learning experience perspective. Therefore, we integrate a score computed via our GEC tools into the updated [completion metric](Implementation) of the discussion task. Finally, our integration of GEC tools can potentially be of use in other parts of the app in the future.
```

So our goal is to provide a score to the group to decide whether they pass or fail the tasks. And in doing so we want all users to contribute equally and as we provide general feedback will the most common mistakes that will be beneficial to all the participants to know about those mistakes and avoid them. In this way the core of the task as group progress will be intact, and each individual will also get benefit from this.
The discussion task will make more sense if the user write a sentence that is related to the topic that is given by the bot to discuss. But the scope of our NLP implementation will only be limited to the correctness of the sentences in grammatical sense and the amount of participation. In order to make sure that users are discussing a related topic as given by the bot, we could use an NLP topic model to check how closely the given topic and the sentences written by users are related. Based on the closeness of the topic and the sentences, we can add extra scores to make sure the discussion stays on topic. This is not implemented in the current version, but is on possible improvement in future versions of the app.

## Tools and Models

### LanguageTool

> The first tool we tested and evaluated for our task is the python library `language_tool_python` (https://pypi.org/project/language-tool-python/) which is a Python wrapper for LanguageTool. LanguageTool (LT) is an open-source spelling and grammar checker used in OpenOffice (see the project home page at https://languagetool.org for more information and source code). It was originally developed by Daniel Naber during his diploma work and is currently maintained by him and an open source community (Naber, 2003). 

Reviewer: ```suggestion
The first tool we tested and evaluated for our task is the python library `language_tool_python` ([https://pypi.org/project/language-tool-python/](https://pypi.org/project/language-tool-python/)) which is a Python wrapper for LanguageTool. LanguageTool (LT) is an open-source spelling and grammar checker used in OpenOffice (see the project home page at https://languagetool.org for more information and source code). It was originally developed by Daniel Naber during his diploma work and is currently maintained by him and an open source community (Naber, 2003). 
```

For example: 

```
text = 'A sentence with an error in the Hitchhiker’s Guide tot he Galaxy'
matches = tool.check(text)
print(matches[1])

>>> Line 1, column 51, Rule ID: TOT_HE[1]
>>> Message: Did you mean 'to the'?
>>> Suggestion: to the
```

The tool can also be used to automatically correct input sentences: 

```
tool.correct(text)
>>> 'A sentence with an error in the Hitchhiker’s Guide to the Galaxy'
```

However, as we tested this feature, we concluded that it can oftentimes be misleading to English learners and propose corrections which do not match the intended message the user wanted to write. Thus, we only use LT to identify errors and track which errors have been made repeatedly. To this end, we identified a set of 20 grammatical error types which we assume new English learners at our target proficiency level will likely make and which we are also fairly certain that LT will detect (see our dataset with examples for each of these errors [here](https://docs.google.com/spreadsheets/d/1RG3Hkb59h8DgtXzJnkh9fbOq3yHjRm2AN00qlpnhlUs/edit#gid=1946205687)). 

> To assure this, we ran an evaluation script and tested the tool on three datasets (see [Fine-tuning and datasets](Fine-tuning and datasets)). 

Reviewer: ```suggestion
To assure this, we ran an evaluation script and tested the tool on three datasets (see [Fine-tuning and Datasets](#fine-tuning-and-datasets)). 
```

The first evaluation step was performed using a corpus constructed from the first training data split from the WI+LOCNESS BEA 2019 corpus (Bryant et al., 2019). That corpus was initially chosen because it consists of sentences (N = 1300) from essays written by English learners close to our target proficiency level. Then, sentences containing our target grammatical structures (see above) were extracted (N = 320). The initial evaluation on this dataset indicated that LT has a lower accuracy than GECToR but also a lower False Positive rate. However, it became clear that it is difficult to assess the performance of the tools with respect to the important error types on this data because most sentences contained several hard-to-disentangle errors. Therefore, we evaluated the tools on two further datasets. 

> We used the dataset created for the [sentence correction task](link) specifically to assess the False Positive rate of the tools since a hand-checked ground truth was given for these sentences. The results of this evaluation can be seen below:

Reviewer: ```suggestion
We used the dataset created for the [sentence correction task]({{ "" | absolute_url }}/2021/03/03/sentence-correction.html) specifically to assess the False Positive rate of the tools since a hand-checked ground truth was given for these sentences. The results of this evaluation can be seen below:
```

**LanguageTool:**    

Classification  |  Counts
:-------------  |  :----------: |
True Positives  |  768
True Negatives  |  1468
False Positives |  33
False Negatives |  733

Accuracy: 0.745  
FPR: 0.022

**GECToR:**

Classification  |  Counts
:-------------  |  :----------: |
True Positives  |  1065
True Negatives  |  1414
False Positives |  95
False Negatives |  444
  
Accuracy: 0.821  
FPR: 0.063

Given the close match in performance of LT and GECToR, we decided to hand-craft a third dataset containing 5-10 examples per error type that we identified (see above). The dataset consists of 237 sentences with 20 different error codes (see our [custom dataset](https://docs.google.com/spreadsheets/d/1RG3Hkb59h8DgtXzJnkh9fbOq3yHjRm2AN00qlpnhlUs/edit?usp=sharing) )

> ![](../assets/images/acc-both-final.png)

Reviewer: ```suggestion
<img src="{{ "" | relative_url }}/assets/images/acc-both-final.png" alt="" class="center">
```

For the quantitative aspect, it is evaluated if every user took part in the discussion based on the individual user participation, which is the total number of words written in this round of discussion. If not every user has met a certain threshold, the bot will send a message to encourage the group to write more in the next round. This threshold is currently set to 15 words per user, but should be fine-tuned and adapted after further testing and evaluation. 

For the qualitative aspect of the feedback, it is checked if the users made many grammatical mistakes in relation to the amount they contributed. It is evaluated whether each user reached a minimum ratio of correct sentences, which is calculated as the quotient between the user score and user participation. If not every user has met a certain ratio, they will be advised to pay more attention to writing correct sentences. This threshold is currently set to 0.5 but should be fine-tuned and adapted after further testing and evaluation. 

> ![](../assets/images/intermediate_feedback_screenshot.jpg)

Reviewer: ```suggestion
<img src="{{ "" | relative_url }}/assets/images/intermediate_feedback_screenshot.jpg" alt="" class="center">
```

At the end of the discussion, a final feedback is sent to the group. The final feedback informs the users which error was most often identified during the discussion (if there were any). It also provides the respective grammatical rule and suggests how to avoid this error in the future. Only common errors which LT can reliably detect are considered here, as we use the error ID provided by the tool to track the number of errors. For each of these common errors, we included a specific feedback message. To ensure that we do not provide feedback for an error which was actually a misclassification (False Positive), the feedback is only sent if the error was detected at least 3 times throughout the discussion. 

> ![](../assets/images/final_feedback_screenshot.jpg)

Reviewer: ```suggestion
<img src="{{ "" | relative_url }}/assets/images/final_feedback_screenshot.jpg" alt="" class="center">
```

> The group passes the task if they meet two requirements. First, each user has to reach a minimum user score in at least two of the three discussion rounds, which is currently set to 5. The second condition is that the group reaches a group threshold together, which is divided into three different levels. The group passes the task at an intermediate level if the score is 5-10; at a good level if the score is 10-15; and at a very good level if the score is at least 15. These conditions are checked in the `is_correct` function in `discussion.py`. The function returns the group level which aligns with the newly developed [storytelling features](storytelling) which requires each task to provide one failure and three possible pass stages. 

Reviewer: ```suggestion
The group passes the task if they meet two requirements. First, each user has to reach a minimum user score in at least two of the three discussion rounds, which is currently set to 5. The second condition is that the group reaches a group threshold together, which is divided into three different levels. The group passes the task at an intermediate level if the score is 5-10; at a good level if the score is 10-15; and at a very good level if the score is at least 15. These conditions are checked in the `is_correct` function in `discussion.py`. The function returns the group level which aligns with the newly developed [storytelling features]({{ "" | absolute_url }}/2021/01/03/storytelling.html) which requires each task to provide one failure and three possible pass stages. 
```

Correctness scoring is also currently flawed. First, using non-words or neologisms does not automatically result in a correctness score of 0. Improving scoring to more confidently punish entirely incorrect sentences will be a more crucial component once the game targets an audience with a higher English skill level. Second, some non-English names are considered typos and result in a correctness score of 0.5 rather than 1, even when the entire sentence is technically correct.

Overall, we improved the evaluation of the discussion task and added automatic GEC to the bot. The grammatical feedback provided to the users is hopefully a valuable support to their learning experience.

## References

Bryant, C., Felice, M., Andersen., Ø. E., and Briscoe, T. (2019). The BEA-2019 Shared Task on Grammatical Error Correction. In Proceedings of the 14th Workshop on Innovative Use of NLP for Building Educational Applications, Florence, Italy. Association for Computational Linguistics.

> Naber, D. (2003). A rule-based style and grammar checker (Diplomarbeit, Universität Bielefeld). http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.6882&rep=rep1&type=pdf

Reviewer: ```suggestion
Naber, D. (2003). A rule-based style and grammar checker (Diplomarbeit, Universität Bielefeld). [http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.6882&rep=rep1&type=pdf](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.148.6882&rep=rep1&type=pdf)
```

	- [Fine-tuning and Datasets](#fine-tuning-and-datasets)
4. [Implementation](#implementation)
	- [Grammatical Error Detection Pipeline](#grammatical-error-detection-pipeline)
	- [Grammar Feedback and Passing the Task](#grammar-feedback-and-passing-the-task)
5. [Outlook and Limitations](#outlook-and-limitations)
6. [References](#references)

## Motivation

Learning to correctly apply grammatical structures in written language is an important and often difficult task for English learners. Therefore, during this semester, grammatical error correction (GEC) was introduced in Escapeling in order to help learners improve their grammar skills. More specifically, state-of-the-art [NLP-tools for GEC](#tools-and-models) were employed in the [discussion task]({{ "" | absolute_url }}/2021/03/02/discussion.html).

The discussion task offers a suitable framework for actively learning and practicing grammar because in this task, users actively produce language, as opposed to focusing on language comprehension in [other tasks]({{ "" | absolute_url }}/#three). To that end, we include diverse feedback on the users’ contributions both after each round of the discussion and at the end of the task.

> 

Reviewer 2: > the users’ contributions

I am not sure if this really is wrong, but it kinda looks unusual… I tried DeepL which suggested:
on user contributions

Another reason for adding GEC to the discussion task is the improvement of the completion metric for this task. That is, in line with the [escape-room]({{ "" | absolute_url }}/2021/01/03/storytelling.html) framing of the app, users’ performance is linked to getting pieces of the code in order to complete the mission. Previously, the performance was only evaluated quantitatively, i.e., by setting a threshold on the minimal number of words users need to produce in order to pass the task. We deem that adding a qualitative evaluation of the contributions to the completion metric improves the accuracy and is advantageous from a learning experience perspective. Therefore, we integrate a score computed via our GEC tools into the updated [completion metric]({{ "" | absolute_url }}/2021/03/02/discussion#design) of the discussion task. Finally, our integration of GEC tools can potentially be of use in other parts of the app in the future.

> 

Reviewer 2: > users’ performance

It also looks kinda unusual, I would write something like:
the performance of users 
The user performance
the users performance

Below, details can be found on how we operationalize GEC in order to improve the learning experience offered to our users in the discussion task.

## Goal and Scope

> 

Reviewer 3: •	the “Goal and Scope” section is overall nicely written, giving details and clear justifications for the decisions taken along the way, limitations, possible extensions of the implementation – well done! 😊
•	a small stylistic suggestion: sometimes the same words occur a lot in a relatively short text span (eg. “users” in the first paragraph), which makes the text a bit cumbersome. Maybe try using more synonyms, pronouns or passives to say the same message with a more variety of words, it makes the text more enjoyable for the brain to read 😊 
•	it may be a good idea to give a little more detail on why/how specifically the types of mistakes, which are to be tracked by the model, were picked. Also, in this section only 5 types of mistakes are listed, but in the implementation, a list of 20 is provided. Are these just some examples? Or have they later been narrowed down to just these 5? It isn’t clear to me from the text.
•	section within lines 48-52 seems a bit repetitive. I think it would also make sense to change the order of the paragraph/main points a bit – to start with line 50-51 which gives reasons for picking only a few types of mistakes, then follow by 48-49 which explains what feedback is given to these mistakes. This is also the order which you use in the previous sentence and I’d say it is more intuitive to stick to it rather than jump back and forth.

The idea behind this is to make sure that users don’t get an overwhelming correction or incorrect suggestions. As we noticed that model don’t always provide a correct suggestion for the grammatical sentence, we don’t want the user to get confused with the wrong suggestions. More details on this can be found in the following sections. Therefore, at the end of the task, the bot will provide the following feedback regarding grammatical mistakes: "I noticed some recurring mistakes. You can take a look at the most important grammatical rule below:". This is followed by error-specific feedback; for instance, this text will be sent regarding subject-verb agreement: "The subject and the verb have to agree with respect to the number. In present tense, an '-s' ending is added to the verb when the subject of the sentence is singular."

> If no common mistakes were made, the bot provides the following feedback at the end of the task: "I did not notice any recurring gramamtical mistakes during the discussion. Good job!"

Reviewer 2: > that model don’t always provide

doesn't

> any recurring gramamtical mistakes

grammatical

So our goal is to provide a score to the group to decide whether they pass or fail the task. And in doing so we want all users to contribute equally, and as we provide general feedback on the most common mistakes, it will be beneficial to all the participants to know about those mistakes and avoid them. In this way the core of the task as group progress will be intact, and each individual will also benefit from this.
The discussion task would make more sense if the users write sentences that are related to the topic that is given by the bot to discuss. But the scope of our NLP implementation is limited to the correctness of the sentences in grammatical sense and the amount of participation for this semester. In order to make sure that users are discussing a related topic as given by the bot, we could use an NLP topic model to check how closely the given topic and the sentences written by users are related. Based on the closeness of the topic and the sentences, we can add extra scores to make sure the discussion stays on topic. This is not implemented in the current version, but is a possible improvement in future versions of the app.

> 

Reviewer 2: > in grammatical sense

I think 'in a grammatical sense' would be correct

```

However, as we tested this feature, we concluded that it can oftentimes be misleading to English learners and propose corrections which do not match the intended message the user wanted to write. Thus, we only use LT to identify errors and track which errors have been made repeatedly. To this end, we identified a set of 20 grammatical error types which we assume new English learners at our target proficiency level will likely make and which we are also fairly certain that LT will detect (see our dataset with examples for each of these errors [here](https://docs.google.com/spreadsheets/d/1RG3Hkb59h8DgtXzJnkh9fbOq3yHjRm2AN00qlpnhlUs/edit#gid=1946205687)).
To assure this, we ran an evaluation script and tested the tool on three datasets (see [Fine-tuning and datasets](#fine-tuning-and-datasets)).

### GECToR

We decided to chain two models and additionally use a neural transformer-based model GECToR, an open-source pretrained model published by grammarly (Omelianchuk et al., 2020), because LT was quite reliable at catching certain types of errors, but missed other kinds of errors, resulting in a lower accuracy.

> 

Reviewer 2: > We decided to chain two models 

which models? Why did you decide to chain them? I feel a little overwhelmed with the first sentence. 

Maybe changing the order and describing the problem first and then elaborating on GECToR could help:
As described above, LT is quite reliable at catching certain types of errors, but misses other kinds of errors, resulting in a lower accuracy at detecting all errors. In order to solve this problem, we are using two different …models…….
In order to achieve a higher error detection accuracy, we decided to additionally use GECToR, an open-source pretrained model published by grammarly (Omelianchuk et al., 2020). It consists of….

Maybe it also makes sense to only describe GECToR here and put the part where you describe the problem with LT at the beginning of “Fine-tuning and Datasets, because you already have a wrap up sentence there which models you choose and why.

The model consists of a transformer-based sequence tagging encoder with two linear layers on top (Vaswani et al., 2017). The encoder applies error corrections by making token-level transformations of the erroneous input text. To this end, the training data was tagged with transformations minimizing the Levenshtein distance between input and target sequences. The encoder was pretrained on several datasets and is [available on GitHub](https://github.com/grammarly/gector). The linear layers perform error detection and error tagging. Based on our testing results, we chose to use the pretrained RoBERTa backbone for the encoder (Omelianchuk et al., 2020). When using available inference scripts, the model outputs a suggested correct sentence, given an input sentence. That is, off-the-shelf GECToR is not ideal for identifying the types of errors made by users because no error IDs are available. Using the model requires storing the frozen checkpoints (~500MB) and performing inference via a local API (~3GB).

> 

Reviewer 2: > The model consists of a transformer-based sequence tagging encoder with two linear layers on top (Vaswani et al., 2017). The encoder applies error corrections by making token-level transformations of the erroneous input text.

I personally would have some questions about this part because I do not understand all the terms that are used but I think that Thelen is familiar with these terms 😊

> That is, off-the-shelf GECToR is not ideal for identifying the types of errors made by users because no error IDs are available. Using the model requires storing the frozen checkpoints (~500MB) and performing inference via a local API (~3GB).

Then why are you using it? What is GECToR good at?

### Fine-tuning and Datasets

Due to open-source availability and computational cost considerations, LT and GECToR remained as the two final candidate tools for our use case, among different approaches to GEC proposed by the NLP community (e.g., Ng et al., 2014) .

> 

Reviewer 2: >  LT and GECToR remained as the two final candidate tools for our use case, among different approaches to GEC proposed by the NLP community

I thought you were using two different models + GECToR. So one of these models is LT and which one is the other one? Maybe I am confusing models and tools.

Reviewer 4: Thanks for pointing out the confusion. We use two tools (= models)  in total, i.e., LT + GECToR. We will definitely try to make the text more clear regarding that.
